src/Lean/Util/TermUnsafe.lean

```diff
@@ -0,0 +1,62 @@
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+import Lean.Elab.ElabRules
+import Lean.Meta.Closure
+import Lean.Compiler.ImplementedByAttr
+
+/-!
+Defines term syntax to call unsafe functions.
+
+```
+def cool :=
+  unsafe (unsafeCast () : Nat)
+
+#eval cool
+```
+-/
+
+namespace Lean.TermUnsafe
+
+open Lean Meta Elab Term
+
+/-- Construct an auxiliary name based on the current declaration name and the given `hint` base. -/
+def mkAuxName (hint : Name) : TermElabM Name :=
+  withFreshMacroScope do
+    let name := (← getDeclName?).getD Name.anonymous ++ hint
+    pure $ addMacroScope (← getMainModule) name (← getCurrMacroScope)
+
+/--
+`unsafe t : α` is an expression constructor which allows using unsafe declarations inside the
+body of `t : α`, by creating an auxiliary definition containing `t` and using `implementedBy` to
+wrap it in a safe interface. It is required that `α` is nonempty for this to be sound,
+but even beyond that, an `unsafe` block should be carefully inspected for memory safety because
+the compiler is unable to guarantee the safety of the operation.
+
+For example, the `evalExpr` function is unsafe, because the compiler cannot guarantee that when
+you call ```evalExpr Foo ``Foo e``` that the type `Foo` corresponds to the name `Foo`, but in a
+particular use case, we can ensure this, so `unsafe (evalExpr Foo ``Foo e)` is a correct usage.
+-/
+elab "unsafe " t:term : term <= expectedType => do
```

Comment:
Syntax decl and docstring should probably be moved into Init?

Reply:
I moved the syntax and docstring to Init/Notation.lean (is that correct?)

I changed the elab statement to use `builtin_term_elab`.